src/analysis/normed_space/SemiNormedGroup.lean

```diff
@@ -0,0 +1,127 @@
+
+/-- Construct a bundled `SemiNormedGroup` from the underlying type and typeclass. -/
+def of (M : Type u) [semi_normed_group M] : SemiNormedGroup := bundled.of M
+
+instance (M : SemiNormedGroup) : semi_normed_group M := M.str
+
+@[simp] lemma coe_of (V : Type u) [semi_normed_group V] : (SemiNormedGroup.of V : Type u) = V := rfl
+@[simp] lemma coe_id (V : SemiNormedGroup) : ⇑(𝟙 V) = id := rfl
+
+instance : has_zero SemiNormedGroup := ⟨of punit⟩
+instance : inhabited SemiNormedGroup := ⟨0⟩
+
+instance : limits.has_zero_morphisms.{u (u+1)} SemiNormedGroup := {}
+
+end SemiNormedGroup
+
+/--
+`SemiNormedGroup₁` is a type synonym for `SemiNormedGroup`,
+which we shall equip with the category structure consisting only of the norm non-increasing maps.
+-/
+@[derive has_coe_to_sort]
+def SemiNormedGroup₁ : Type (u+1) := bundled semi_normed_group
+
+namespace SemiNormedGroup₁
+
+instance : large_category.{u} SemiNormedGroup₁ :=
+{ hom := λ X Y, { f : normed_group_hom X Y // f.norm_noninc },
+  id := λ X, ⟨normed_group_hom.id, normed_group_hom.norm_noninc.id⟩,
+  comp := λ X Y Z f g, ⟨g.1.comp f.1, g.2.comp f.2⟩, }
+
+@[ext] lemma hom_ext {M N : SemiNormedGroup₁} (f g : M ⟶ N) (w : (f : M → N) = (g : M → N)) :
+  f = g :=
+subtype.eq (normed_group_hom.ext (congr_fun w))
+
+@[simp] lemma val_coe {M N : SemiNormedGroup₁} (f : M ⟶ N) : (f.val : M → N) = (f : M → N) := rfl
+
+instance : concrete_category.{u} SemiNormedGroup₁ :=
+{ forget :=
+  { obj := λ X, X,
+    map := λ X Y f, f, },
+  forget_faithful := {} }
+
+/-- Construct a bundled `SemiNormedGroup₁` from the underlying type and typeclass. -/
+def of (M : Type u) [semi_normed_group M] : SemiNormedGroup₁ := bundled.of M
+
+instance (M : SemiNormedGroup₁) : semi_normed_group M := M.str
+
+/-- Promote a morphism in `SemiNormedGroup` to a morphism in `SemiNormedGroup₁`. -/
+def mk_hom {M N : SemiNormedGroup} (f : M ⟶ N) (i : f.norm_noninc) :
+  SemiNormedGroup₁.of M ⟶ SemiNormedGroup₁.of N :=
+⟨f, i⟩
+
+@[simp] lemma mk_hom_apply {M N : SemiNormedGroup} (f : M ⟶ N) (i : f.norm_noninc) (x) :
+  mk_hom f i x = f x := rfl
+
+/-- Promote an isomorphism in `SemiNormedGroup` to an isomorphism in `SemiNormedGroup₁`. -/
+@[simps]
+def mk_iso {M N : SemiNormedGroup} (f : M ≅ N) (i : f.hom.norm_noninc) (i' : f.inv.norm_noninc) :
+  SemiNormedGroup₁.of M ≅ SemiNormedGroup₁.of N :=
+{ hom := mk_hom f.hom i,
+  inv := mk_hom f.inv i',
+  hom_inv_id' := by { apply subtype.eq, exact f.hom_inv_id, },
+  inv_hom_id' := by { apply subtype.eq, exact f.inv_hom_id, }, }
+
+instance : has_forget₂ SemiNormedGroup₁ SemiNormedGroup :=
+{ forget₂ :=
+  { obj := λ X, X,
+    map := λ X Y f, f.1, }, }
+
+@[simp] lemma coe_of (V : Type u) [semi_normed_group V] : (SemiNormedGroup₁.of V : Type u) = V :=
+rfl
+@[simp] lemma coe_id (V : SemiNormedGroup₁) : ⇑(𝟙 V) = id := rfl
+@[simp] lemma coe_comp {M N K : SemiNormedGroup₁} (f : M ⟶ N) (g : N ⟶ K) :
+  ((f ≫ g) : M → K) = g ∘ f := rfl
```

Comment:
Can't `concrete_category` take care of these?

Reply:
The `zero_comp'` field of `has_zero_morphisms` is sad without these.

Reply:
I see. So really we should make these simp lemmas for arbitrary concrete categories, I guess. (Currently only the fully applied versions are simp lemmas, I think.)

Reply:
But we can do that in a separate PR. This shouldn't wait for it.